app/v2/controllers/V2DeparturesController.scala

```diff
@@ -112,12 +107,16 @@ class V2DeparturesControllerImpl @Inject() (
         implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
         withTemporaryFile {
           (temporaryFile, source) =>
-            validationService
-              .validateXML(MessageType.DepartureDeclaration, source)
-              .leftMap(translateValidationError)
+            for {
+              _ <- validationService.validateXML(MessageType.DepartureDeclaration, source).convertError
+
+              fileSource = FileIO.fromPath(temporaryFile)
+
+              declarationResult <- departuresPersistenceService.saveDeclaration(request.eoriNumber, fileSource).convertError
+            } yield declarationResult
```

Comment:
The use of .convertError here reads a little code-smelly to me so I did an investigation as to why (though not a fix)

I think this is a consequence of BaseError and withTemporaryFile being entangled. Since BaseError is being used as a PresentationError, it effectively forces you to consider the presentation of the error message as a result of using the temporary file, which I'm not entirely sure is a great idea. I think we need a longer look as to how we're structuring these error types to tidy them up, and ensure that the code we're writing is actually keep seperate concerns and not tying our hands unhelpfully (I think that might be the case with withTemporaryFile atm, because of what BaseError is responsible for)

Reply:
> I think this is a consequence of BaseError and withTemporaryFile being entangled.

It's not, `withTemporaryFile` just conformed to the type required by the for comprhension - and I've made that detachment.

The original issue still stands. The problem is that, ideally, I wanted to use such a structure to make the flatmap chain with `EitherT`s quite simple -- though you may recall there were leftMaps on each call. The problem is we need to be able to unify the various service errors in some way - this basically does the conversion into a presentation error so that the left types of the EitherTs remain consistent. Ultimately, we need to do the type unification somewhere. I imagine that's why David just had his `TransitMovementError` class that basically took the role of _all_ error handling.

If we were to introduce the notion of a base service error type, such as `ServiceError`, that's fine, but we then have to make all services return `EitherT[Future, ServiceError, _]` thanks to generic invarience - and I'd argue that's just as unhelpful as what we're doing here because the service layer is now being dictated by what the presentation layer requires, where here, the presentation layer concern is done in this presentation layer.

The reality is that we could just be absolutely overdoing this. These are microservices - are we overarchitecting to the point where we're just tying ourselves in knots trying to make things look nice and enforce a very clean separation? Something has to give. We just need to decide what.